File: examples_v2/example_sdk_typescript/data_classes/categoryNode.ts
/**
 * Data classes for Category Node.
 *
 * @packageDocumentation
 */

import type {
  NodeInstance, NodeInstanceWrite,
} from "../instance_api/types/instance.ts";
import { InstanceList } from "../instance_api/types/instance.ts";
import type { ViewReference } from "../instance_api/types/references.ts";
import {
  FilterContainer,
  TextFilter,
} from "../instance_api/types/dtypeFilters.ts";

/** View reference for CategoryNode */
export const CATEGORY_NODE_VIEW: ViewReference = {
  space: "pygen_example",
  externalId: "CategoryNode",
  version: "v1",
};

/**
 * Write class for Category Node instances.
 *
 * View for category nodes with reverse direct relation
 */
export interface CategoryNodeWrite extends NodeInstanceWrite {
  readonly instanceType: "node";
  categoryName: string;
}

/**
 * Read class for Category Node instances.
 *
 * Contains all properties including system metadata.
 */
export interface CategoryNode extends NodeInstance {
  readonly instanceType: "node";
  readonly categoryName: string;
}

/**
 * Converts a CategoryNode read instance to a write instance.
 *
 * @param instance - The CategoryNode to convert
 * @returns A CategoryNodeWrite instance
 */
export function categoryNodeAsWrite(instance: CategoryNode): CategoryNodeWrite {
  const { dataRecord, ...rest } = instance;
  const write: CategoryNodeWrite = {
    ...rest,
  };
  if (dataRecord) {
    (write as { dataRecord?: { existingVersion: number } }).dataRecord = {
      existingVersion: dataRecord.version,
    };
  }
  return write;
}

/**
 * List of Category Node instances.
 */
export class CategoryNodeList extends InstanceList<CategoryNode> {
  /**
   * Creates a new CategoryNodeList.
   *
   * @param items - Initial items to populate the list
   */
  constructor(items?: readonly CategoryNode[]) {
    super(items, CATEGORY_NODE_VIEW);
  }

  /**
   * Converts all instances in the list to write instances.
   *
   * @returns Array of CategoryNodeWrite instances
   */
  asWrite(): CategoryNodeWrite[] {
    return this.map(categoryNodeAsWrite);
  }

  /**
   * Creates a new CategoryNodeList from an array.
   *
   * @param items - Array of CategoryNode instances
   * @returns A new CategoryNodeList
   */
  static fromArray(items: readonly CategoryNode[]): CategoryNodeList {
    return new CategoryNodeList(items);
  }
}

/**
 * Filter container for Category Node instances.
 *
 * Provides type-safe filters for all CategoryNode properties.
 */
export class CategoryNodeFilter extends FilterContainer {
  /** Filter for the categoryName property */
  readonly categoryName: TextFilter;

  /**
   * Creates a new CategoryNodeFilter.
   *
   * @param operator - How to combine filters ("and" or "or"). Defaults to "and"
   */
  constructor(operator: "and" | "or" = "and") {
    const categoryName = new TextFilter(CATEGORY_NODE_VIEW, "category_name", operator);

    super([categoryName], operator, "node");

    this.categoryName = categoryName;
  }
}

File: examples_v2/example_sdk_python/data_classes/category_node.py
from typing import ClassVar, Literal

from cognite.pygen._python.instance_api.models._references import ViewReference
from cognite.pygen._python.instance_api.models.dtype_filters import (
    FilterContainer,
    TextFilter,
)
from cognite.pygen._python.instance_api.models.instance import (
    Instance,
    InstanceList,
    InstanceWrite,
)


class CategoryNodeWrite(InstanceWrite):
    """Write class for Category Node instances."""

    _view_id: ClassVar[ViewReference] = ViewReference(space="pygen_example", external_id="CategoryNode", version="v1")
    instance_type: Literal["node"] = Field("node", alias="instanceType")
    category_name: str


class CategoryNode(Instance):
    """Read class for Category Node instances."""

    _view_id: ClassVar[ViewReference] = ViewReference(space="pygen_example", external_id="CategoryNode", version="v1")
    instance_type: Literal["node"] = Field("node", alias="instanceType")
    category_name: str

    def as_write(self) -> CategoryNodeWrite:
        """Convert to write representation."""
        return CategoryNodeWrite.model_validate(self.model_dump(by_alias=True))


class CategoryNodeList(InstanceList[CategoryNode]):
    """List of Category Node instances."""

    _INSTANCE: ClassVar[type[CategoryNode]] = CategoryNode


class CategoryNodeFilter(FilterContainer):
    def __init__(self, operator: Literal["and", "or"] = "and") -> None:
        view_id = CategoryNode._view_id
        self.category_name = TextFilter(view_id, "category_name", operator)
        super().__init__(
            data_type_filters=[
                self.category_name,
            ],
            operator=operator,
            instance_type="node",
        )
